File: backend/app/schemas/mapping.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from app.models.mapping import MappingStatus


class TransformBase(BaseModel):
    order: int
    fn: str
    params: Optional[Dict[str, Any]] = None


class TransformResponse(TransformBase):
    id: int
    mapping_id: int

    class Config:
        from_attributes = True


class SuggestionResponse(BaseModel):
    id: int
    mapping_id: int
    candidates: List[Dict[str, Any]]

    class Config:
        from_attributes = True


class SelectionOption(BaseModel):
    value: str
    label: str


class CustomFieldDefinition(BaseModel):
    technical_name: str
    field_label: str
    field_type: str  # Char, Integer, Float, Boolean, Date, Datetime, Text, Html, Selection, Many2one, Monetary
    required: bool = False
    size: Optional[int] = None  # For Char fields
    help_text: Optional[str] = None
    selection_options: Optional[List[SelectionOption]] = None  # For Selection fields
    related_model: Optional[str] = None  # For Many2one fields


class FieldTypeSuggestion(BaseModel):
    field_type: str
    suggested_size: Optional[int] = None
    selection_options: Optional[List[Dict[str, str]]] = None
    required: bool = False
    rationale: str


class MappingBase(BaseModel):
    header_name: str
    target_model: Optional[str] = None
    target_field: Optional[str] = None
    confidence: Optional[float] = None
    status: MappingStatus = MappingStatus.PENDING
    rationale: Optional[str] = None


class MappingUpdate(BaseModel):
    target_model: Optional[str] = None
    target_field: Optional[str] = None
    status: Optional[MappingStatus] = None
    chosen: Optional[bool] = None
    custom_field_definition: Optional[CustomFieldDefinition] = None


class MappingResponse(MappingBase):
    id: int
    dataset_id: int
    sheet_id: int
    chosen: bool
    transforms: List[TransformResponse] = []
    suggestions: List[SuggestionResponse] = []
    custom_field_definition: Optional[Dict[str, Any]] = None

    class Config:
        from_attributes = True


class MappingListResponse(BaseModel):
    mappings: List[MappingResponse]
    total: int

File: frontend/src/types/mapping.ts
export interface Candidate {
  model: string
  field: string
  confidence: number
  method: string
  rationale: string
}

export interface Suggestion {
  id: number
  mapping_id: number
  candidates: Candidate[]
}

export interface SelectionOption {
  value: string
  label: string
}

export interface CustomFieldDefinition {
  technical_name: string
  field_label: string
  field_type: string
  target_model?: string
  required: boolean
  size?: number
  help_text?: string
  selection_options?: SelectionOption[]
  related_model?: string
}

export interface Transform {
  id: number
  mapping_id: number
  order: number
  fn: string
  params?: Record<string, any> | null
}

export interface Mapping {
  id: number
  dataset_id: number
  sheet_id: number
  header_name: string
  target_model: string | null
  target_field: string | null
  confidence: number | null
  status: 'pending' | 'confirmed' | 'ignored' | 'create_field'
  chosen: boolean
  rationale: string | null
  suggestions: Suggestion[]
  transforms?: Transform[]
  custom_field_definition?: CustomFieldDefinition
  lambda_function?: string
}

export interface FieldTypeSuggestion {
  field_type: string
  suggested_size?: number
  selection_options?: Array<{ value: string; label: string }>
  required: boolean
  rationale: string
}

export interface ColumnProfile {
  id: number
  name: string
  dtype_guess: string
  null_pct: number
  distinct_pct: number
  patterns: Record<string, number>
  sample_values: string[]
}

export interface Dataset {
  id: number
  name: string
  source_file_id: number
  created_at: string
  sheets: Sheet[]
}

export interface Sheet {
  id: number
  dataset_id: number
  name: string
  n_rows: number
  n_cols: number
}
